src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp

```diff
@@ -265,59 +369,32 @@ void ov::npuw::LLMInferRequest::infer_generate(ov::SoPtr<ov::ITensor> input_ids,
-                LOG_DEBUG("Input name " << input_name << " doesn't contain kv cache. Skipping.");
-                continue;
-            }
-
-            auto kvcache_in_tensor = m_kvcache_request->get_tensor(m_kvcache_in_ports.at(input_name));
-
-            // FIXME: We don't need to fill whole tensor with 0s, but only tensor.size() - num_stored_tokens
-            //        taking into account kvcache dimension.
-            fill_tensor<ov::float16>(kvcache_in_tensor, 0);
+    // TODO: Use another solution than flag if multithreading comes to play.
+    if (!m_generate_initialized) {
+        LOG_DEBUG("Copy kv-cache asynchronously...");
+        auto copy_kvcache_future = std::async(std::launch::async, &LLMInferRequest::copy_kvcache, this);
 
-            const auto& kv_dim = (output_name.find("value") != std::string::npos && kvcache_desc.v_tensors_transposed)
-                                     ? 3u
-                                     : kvcache_desc.dim;
-
-            auto prefill_out_slice = make_tensor_slice(prefill_out_tensor,
-                                                       kv_dim,
-                                                       kvcache_desc.max_prompt_size - kvcache_desc.num_stored_tokens,
-                                                       kvcache_desc.max_prompt_size);
-
-            auto kvcache_in_slice = make_tensor_slice(kvcache_in_tensor, kv_dim, 0u, kvcache_desc.num_stored_tokens);
-
-            if (kv_dim == 3u) {
-                copy_columns_by_row_chunks(prefill_out_slice, kvcache_in_slice);
-            } else if (kv_dim == 2u) {
-                copy_by_planes(prefill_out_slice, kvcache_in_slice);
-            } else {
-                prefill_out_slice->copy_to(kvcache_in_slice._ptr);
-            }
+        LOG_DEBUG("Prepare attention mask pattern.");
+        auto kv_attn_mask = m_kvcache_request->get_tensor(m_kvcache_in_ports.at("attention_mask"));
+        fill_tensor<int64_t>(kv_attn_mask, 0);
+        // NOTE: Attention mask pattern for generate model requires last "1" to be in the end of the mask.
+        //       We can safely set this "1" once and then copy on one "1" less in the infer_generate().
+        kv_attn_mask->data<int64_t>()[m_npuw_llm_compiled_model->m_kvcache_desc.total_size - 1] = 1;
+        if (m_tail_mm_request) {
+            LOG_DEBUG("Set input tensor for tail inference request from generate model output.");
+            m_tail_mm_request->set_tensor(
+                m_tail_embed_port,
+                m_kvcache_request->get_tensor(m_kvcache_out_ports.at(ov::npuw::LLMCompiledModel::output_embeds)));
         }
 
-        LOG_DEBUG("Prepare attention mask pattern.");
-        auto* attention_mask_data =
-            m_kvcache_request->get_tensor(m_kvcache_in_ports.at("attention_mask"))->data<int64_t>();
-        attention_mask_data[kvcache_desc.total_size - 1] = 1;
+        copy_kvcache_future.wait();
```

Comment:
Was it worth that? I'd expect all the above code to run pretty short to be noticeable? Do you have performance numbers without future here? Would it spawn a new thread every time?